### PR TITLE
test: improve unit tests for container name autocompletion

### LINKED ISSUE #1600: Improve tests for container-name autocompletion for exec commands
### What would you like to be added?

Tests in container-name autocompletion are done in separate functions and they don't cover the actual functionality which is auto-completing the container-name.
We can compact those methods in a test table and add more tests to cover more cases. 

### Why is this needed?

This is needed to improve the confidence in these tests and make them more maintainable and clean.

### PR DESCRIPTION
**Pull request by TessaIO**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
Cleanup and add some tests for unit tests for container name autocompletion.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubectl/issues/1600

#### Does this PR introduce a user-facing change?

```release-note
N/A
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
N/A
```

### CODE REVIEW

**Comment by TessaIO** [comment]
ping @ardaguclu @ah8ad3 since you already have context on this.

File: staging/src/k8s.io/kubectl/pkg/util/completion/completion_test.go [hunk 1]
```diff
@@ -330,57 +330,76 @@ func TestPodResourceNameCompletionFuncJointFormTooManyArgs(t *testing.T) {
...
+func TestResourceAndContainerNameCompletionFunc(t *testing.T) {
+	barPod := getTestPod()
 
-func TestPodResourceNameAndContainerCompletionFuncJointFormTooManyArgs(t *testing.T) {
-	tf, cmd := prepareCompletionTest()
-	pods, _, _ := cmdtesting.TestData()
-	addResourceToFactory(tf, pods)
+	testCases := []struct {
+		name              string
+		obj               runtime.Object
```
Comment (ah8ad3) [inline staging/src/k8s.io/kubectl/pkg/util/completion/completion_test.go:338]: It seems you only using `barPod` as obj, is there any reason to add obj in test table?

**Review by ah8ad3 — COMMENTED** [review]
Thanks for PR!
I haven't check to see if is there any other function that can be replaced with this test table too.

**Comment by TessaIO** [comment]
@ah8ad3 should I update all of them?

**Comment by ah8ad3** [comment]
> should I update all of them?

My bad on not specifying my opinion, i think if there is for example 5 other tests that can be removed and put in the table here it would be better.
Updating all the tests could make reviews harder and maybe take long time to review.

**Comment by TessaIO** [comment]
> > should I update all of them?
> 
> My bad on not specifying my opinion, i think if there is for example 5 other tests that can be removed and put in the table here it would be better. Updating all the tests could make reviews harder and maybe take long time to review.

There are many of them in the `completion_test.go`. I suggest doing that in a separate PR to make the review easier. wdyt?

File: staging/src/k8s.io/kubectl/pkg/util/completion/completion_test.go (hunk 1 above)
Comment (TessaIO) [reply staging/src/k8s.io/kubectl/pkg/util/completion/completion_test.go]: Updated thanks!

**Review by ah8ad3 — COMMENTED** [review]
LGTM thanks!

**Comment by AhmedThresh** [comment]
ping @ardaguclu @brianpursley

**Comment by TessaIO** [comment]
ping @brianpursley @ardaguclu

File: staging/src/k8s.io/kubectl/pkg/util/completion/completion_test.go [hunk 2]
```diff
@@ -330,57 +330,68 @@ func TestPodResourceNameCompletionFuncJointFormTooManyArgs(t *testing.T) {
...
+		},
+		{
+			name:              "no args resources",
+			args:              []string{},
+			toComplete:        "s",
+			expectedComps:     []string{"services/", "statefulsets/"},
+			expectedDirective: cobra.ShellCompDirectiveNoFileComp | cobra.ShellCompDirectiveNoSpace,
+		},
+		{
+			name:              "joint from no args",
```
Comment (brianpursley) [inline staging/src/k8s.io/kubectl/pkg/util/completion/completion_test.go:357]:
```suggestion
			name:              "joint form no args",
```

File: staging/src/k8s.io/kubectl/pkg/util/completion/completion_test.go [hunk 3]
```diff
@@ -330,57 +330,68 @@ func TestPodResourceNameCompletionFuncJointFormTooManyArgs(t *testing.T) {
...
+			expectedDirective: cobra.ShellCompDirectiveNoFileComp,
+		},
+	}
 
-	compFunc := PodResourceNameAndContainerCompletionFunc(tf)
-	comps, directive := compFunc(cmd, []string{"pod/pod-name", "container-name"}, "")
-	checkCompletion(t, comps, []string{}, directive, cobra.ShellCompDirectiveNoFileComp)
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			tf, cmd := prepareCompletionTest()
```
Comment (brianpursley) [inline staging/src/k8s.io/kubectl/pkg/util/completion/completion_test.go:388]: These 3 lines are the same for every test. No need to rerun them for each test. Are you able to move these lines outside the for loop?
```
			tf, cmd := prepareCompletionTest()
			addResourceToFactory(tf, barPod)
			compFunc := PodResourceNameAndContainerCompletionFunc(tf)
```

**Review by brianpursley — CHANGES_REQUESTED** [review]

File: staging/src/k8s.io/kubectl/pkg/util/completion/completion_test.go (hunk 3 above)
Comment (TessaIO) [reply staging/src/k8s.io/kubectl/pkg/util/completion/completion_test.go]: Good catch, thanks!

**Comment by TessaIO** [comment]
@brianpursley thanks for the review! Updated and PTAL.

File: staging/src/k8s.io/kubectl/pkg/util/completion/completion_test.go (hunk 3 above)
Comment (brianpursley) [reply staging/src/k8s.io/kubectl/pkg/util/completion/completion_test.go]: OK, I see the unit tests failed after doing this, so I think you should put it back as it was.  I didn't realize it needed to be inside the loop.

**Comment by brianpursley** [comment]
I see the unit tests failed after moving the lines as I had suggested. I'm not sure why those lines need to be inside the loop, but there must be something that needs to be reset for each test (iostreams maybe?). I think you can put those lines back inside the loop like you had them and it should be fine.

**Comment by TessaIO** [comment]
> I see the unit tests failed after moving the lines as I had suggested. I'm not sure why those lines need to be inside the loop, but there must be something that needs to be reset for each test (iostreams maybe?). I think you can put those lines back inside the loop like you had them and it should be fine.

Thanks and done. Also maybe it's one of the reasons that previous tests were separated.
Anyways PTAL when you have time.

**Review by ah8ad3 — COMMENTED** [review]
Thanks!
/lgtm

**Comment by ardaguclu** [comment]
That looks good to me and thank you. But next time it is better to work on the issues that are labelled with triage/accepted to prevent any waste of effort as request might be rejected by the maintainers.
/approve

**Pull request merged**